### ISSUE TITLE
Getting nan as loss

### ISSUE THREAD

**Issue by simonepri** [facebookresearch/PyTorch-BigGraph#53]
## Observed Results

![image](https://user-images.githubusercontent.com/3505087/58751992-75aa2300-84a7-11e9-94df-1d316438a907.png)

## Expected Results

Loss must be always a number

## Relevant Code
  ```
I will provide a repro code as soon as possible.
  ```

**Comment by lerks** [comment]
We've seen such an issue before but we've never been able to pin it down well enough to understand the root cause. We believe it may be due to divergence occurring during the optimizer step and we've usually been able to mitigate it by fixing some hyperparameters (say, the learning rate). It would be great if we could get a consistently reproducible example that allows us to investigate this better.

I've noticed that this seems to occur more often in a distributed setting, so another factor could be the instability/noise introduced by systems like the parameter server or by approaches like hogwild, which apply asynchronous updates to the parameters and may end up corrupting them.

If you have a small enough setup in which this is happening that is fast enough to iterate on (and you may have it, as you've been able to do 600 epochs on it), you may want to try to deactivate the main sources of nondeterminism and see if you can get a repro. 
- The first of these sources is the on-the-fly shuffling of the edges before training (you could instead just shuffle them once and store them in that order).
- The second one is the hogwild-style updates made by the multiple processes on the same machine. The easiest way to disable it is by just using one worker (hence one processor) but this may slow you down considerably, alternatively you may implement some locking around the parameter updates, but this is more complicated.
- Then there's the uniform sampling of negatives, which you can either choose to replace with a generator with a fixed seed or disable entirely.

Unfortunately there's not much more I am able to recommend to fix or prevent this. If you have any ideas yourself please let us know!

**Comment by simonepri** [comment]
> Unfortunately there's not much more I am able to recommend to fix or prevent this. If you have any ideas yourself please let us know!

What about enabling pytorch's anomaly detection feature by default?
```python
torch.autograd.set_detect_anomaly(True)
```

If we do so, when there's a problem PyTorch gives insightful informations instead of just giving `nan`.
E.G:
![image](https://user-images.githubusercontent.com/3505087/59561122-8b2d5a00-901c-11e9-9c48-fe1867414f60.png)

**Comment by simonepri** [comment]
> We've seen such an issue before but we've never been able to pin it down well enough to understand the root cause.

Do you think could be related [to this](https://github.com/facebookresearch/PyTorch-BigGraph/blob/master/torchbiggraph/model.py#L570)?
![image](https://user-images.githubusercontent.com/3505087/59563337-e66d4580-9038-11e9-99a7-64d5a4670764.png)
